components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1915,6 +1928,56 @@ where
         }
     }
 
+    fn check_unsafe_recovery_state(&mut self) {
+        match &self.fsm.peer.unsafe_recovery_state {
+            Some(UnsafeRecoveryState::WaitApply { .. }) => self
+                .fsm
+                .peer
+                .unsafe_recovery_maybe_finish_wait_apply(/*force=*/ false),
+            Some(UnsafeRecoveryState::DemoteFailedVoters {
+                syncer,
+                failed_voters,
+                commit_index,
+                demote_after_exit,
+            }) => {
+                if self.fsm.peer.raft_group.raft.raft_log.applied >= *commit_index {
+                    if *demote_after_exit {
+                        if !self.fsm.peer.is_force_leader() {
+                            error!(
+                                "Unsafe recovery, lost forced leadership after exiting joint state";
+                                "region" => ?self.region(),
+                            );
+                        }
+                        let syncer_clone = syncer.clone();
+                        let failed_voters_clone = failed_voters.clone();
+                        self.unsafe_recovery_demote_failed_voters(
+                            syncer_clone,
+                            failed_voters_clone,
+                        );
+                    } else {
+                        info!("Unsafe recovery, exiting joint state."; "region" => ?self.region());
```

Comment:
the proposal may fail to apply, check if it's in joint state here

Reply:
And better to check the raft configuration here again to ensure the correctness.

Reply:
Done